### ISSUE TITLE
Raster normals with spheremap produces reversed output from Tangram JS

### ISSUE THREAD

**Issue by blair1618** [tangrams/tangram-es#869]
This is demonstrated in the walkabout style: https://tangrams.github.io/magazine/?scene=https://tangrams.github.io/walkabout-style/walkabout-style.yaml#13/37.7639/-122.4323

<img width="690" alt="screen shot 2016-07-26 at 9 50 54 am" src="https://cloud.githubusercontent.com/assets/3371850/17149796/d89996fe-533a-11e6-874a-efc2563dfb90.png">

In walkabout this effect is implemented with a custom shader: https://github.com/tangrams/walkabout-style/blob/gh-pages/walkabout-style.yaml#L709

Tangram-es injects normals in the same way as JS: https://github.com/tangrams/tangram-es/blob/5d80c7e586dc7c76142547de73dd1a264f518569/core/shaders/polygon.fs#L66

Tangram-es also appears to calculate normals from textures in the same way as JS too: https://github.com/tangrams/tangram/blob/b4a9ec50700c880df272c983acd2ab6e93bc0ef6/src/gl/shaders/material.glsl#L122

**Comment by bcamper** [comment]
It appears that JS and ES are reading the texture with opposite vertical orientations.

It's instantiated as a uniform (absolute path added for testing):

```
       shaders:
            uniforms:
                u_envmap: https://tangrams.github.io/walkabout-style/images/draw-test9.jpg
```

If the style is modified to directly visualize the texture in world coordinates (scene file attached):

`color = texture2D(u_envmap, fract(worldPosition().xy*.00025));`

You can see that JS has the [original image](https://tangrams.github.io/walkabout-style/images/draw-test9.jpg) is "right-side up", while ES has it "upside-down" -- though, note that the UVs are sampling from the bottom-left:

<img width="1116" alt="screen shot 2016-07-26 at 4 30 16 pm" src="https://cloud.githubusercontent.com/assets/16733/17159275/e8799bda-5350-11e6-84c0-7913c3efa313.png">

![tangram-tue jul 26 2016 16-42-11 gmt-0700 pdt](https://cloud.githubusercontent.com/assets/16733/17159280/ef9922e6-5350-11e6-8022-98ea854f2ca7.png)

I'm actually unclear which one is correct :) (even though JS shows the visually desired output... it could be an accident).

**Comment by blair1618** [comment]
tl;dr: textures need to be flipped on tangram-es. 

To quote http://docs.gl/es2/glTexImage2D on the ordering of texture data:

> The first element corresponds to the lower left corner of the texture image. Subsequent elements progress left-to-right through the remaining texels in the lowest row of the texture image, and then in successively higher rows of the texture image. The final element corresponds to the upper right corner of the texture image.

Now compare the documentation for stb_image.h (https://github.com/nothings/stb/blob/master/stb_image.h#L246-L256):

> The return value from an image loader is an 'unsigned char *' which points to the pixel data, or NULL on an allocation failure or if the image is corrupt or invalid.The pixel data consists of *y scanlines of *x pixels, with each pixel consisting of N interleaved 8-bit components; the first pixel pointed to is top-left-most in the image.

and it becomes pretty clear that the data returned from stb_image needs to be flipped before it can be used correctly in OpenGL.

Lucky for us, there's a handy function in stb_image that we can use to change that in one line: https://github.com/nothings/stb/blob/master/stb_image.h#L503

And we even have a code path for using this when we load a texture in tangram-es: https://github.com/tangrams/tangram-es/blob/master/core/src/gl/texture.cpp#L39

So to correct this behavior in tangram-es, we just have to:

1. Use the 'flip-y' switch on all texture loads
2. Update our sprite code to flip pixel coordinates when converting to UVs. 
3. ???
4. Profit

**Comment by bcamper** [comment]
⬆️ 🌄 ⬇️

**Comment by tallytalwar** [comment]
Nice team work debugging this.
